src/Optimizer/Transformer/OptimizeHeroImages.php

```diff
@@ -0,0 +1,709 @@
+            return;
+        }
+
+        $maxHeroImageCount = $this->configuration->get(OptimizeHeroImagesConfiguration::MAX_HERO_IMAGE_COUNT);
+
+        $this->inlineStyleBackupAttribute = $this->configuration->get(
+            OptimizeHeroImagesConfiguration::INLINE_STYLE_BACKUP_ATTRIBUTE
+        );
+
+        $heroImages     = $this->findHeroImages($document, $maxHeroImageCount);
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > $maxHeroImageCount) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = $maxHeroImageCount;
+        }
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->removeLazyLoading($heroImages[$index]);
+            $this->generatePreload($heroImages[$index], $document, $errors);
+            $this->generateImg($heroImages[$index], $document);
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document          Document to look for hero images in.
+     * @param int      $maxHeroImageCount Maximum number of hero images to accept.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document, $maxHeroImageCount)
+    {
+        $heroImages          = [];
+        $heroImageCandidates = [];
+        $heroImageFallback   = null;
+        $seenParagraphCount  = 0;
+        $node                = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof Element) {
+                $node = NodeWalker::nextNode($node);
+                continue;
+            }
+
+            if ($node->tagName === Tag::P) {
+                $seenParagraphCount++;
+            }
+
+            $heroImage = $this->detectImageWithAttribute($node, Attribute::DATA_HERO);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            } elseif (count($heroImageCandidates) < $maxHeroImageCount) {
+                $heroImageCandidate = $this->detectImageWithAttribute($node, Attribute::DATA_HERO_CANDIDATE);
+                if ($heroImageCandidate) {
+                    $heroImageCandidates[] = $heroImageCandidate;
+                } elseif ($seenParagraphCount < 2 && ! $heroImageFallback) {
+                    $heroImageFallback = $this->detectPossibleHeroImageFallbacks($node);
+                }
+            }
+
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = NodeWalker::skipNodeAndChildren($node);
+            } else {
+                $node = NodeWalker::nextNode($node);
+            }
+        }
+
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Since there were no hero images, use the hero image candidates instead.
+        $heroImages = array_slice($heroImageCandidates, 0, $maxHeroImageCount);
```

Comment:
Since `$heroImages` is known to be empty here, it seemed preferable to use `array_slice()` rather than a while loop. See af51a4636e0f9b77f5556d26834d270e9b7fa444.